pocs/sensors/arduino_io.py

```diff
@@ -99,148 +102,163 @@ def open_serial_device(port, serial_config=None, **kwargs):
+        event that the device disconnects from USB.
         """
-        with self._last_reading_lock:
-            return self._last_reading
-
-    def start_reading(self):
-        """Starts a reader thread that reads reports and writes them to a queue."""
-        if self._thread and self._thread.is_alive():
-            self._logger.debug('Thread {!r} is already running with ident {!r}', self._thread.name,
-                               self._thread.ident)
-            return
-
-        def reader():
-            """The function that is executed by the reader thread."""
-            self._logger.info('Started reading from Arduino {!r} on thread with ident {!r}',
-                              threading.current_thread().name, threading.get_ident())
-            announce = False
-            while not self._stop.is_set():
-                reading = self._get_reading()
-                if self._stop.is_set():
-                    break
-                if not reading:
-                    # Consider adding an error counter.
-                    announce = True
-                    continue
-                if announce:
-                    self._logger.info('Succeeded in reading from {!r}; got:\n{!r}', self.port,
-                                      reading)
-                    announce = False
-                self._handle_reading(reading)
-            self._logger.info('Stopping reading from Arduino {!r} on thread with ident {!r}',
-                              threading.current_thread().name, threading.get_ident())
-
-        self._thread = threading.Thread(target=reader, name=self.board_name)
-        self._thread.daemon = True
-        self._thread.start()
-
-    def stop_reading(self):
-        """Stops the reader thread."""
-        if self._thread and self._thread.is_alive():
-            self._logger.info('Waiting for thread {!r} to stop.', self._thread.name)
-            self._stop.set()
-            self._thread.join(timeout=4.0)
-            if self._thread.is_alive():
-                self._logger.error('Thread {!r} is still running.', self._thread.name)
-            else:
-                self._thread = None
-
-    def write(self, text):
-        """Writes text (a string) to the port.
-
-        Returns: the number of bytes written.
-        """
-        with self._serial_data_lock:
-            if not self._serial_data.is_connected:
-                self._serial_data.connect()
-            return self._serial_data.write(text)
+        while self._keep_running:
+            self.read_and_record()
+            self.handle_commands()
+
+    def read_and_record(self):
+        """Read the next reading available reading and write it to a PanDB collection."""
+        reading = self.get_reading()
+        if not reading:
+            # Consider adding an error counter.
+            self._report_next_reading = True
+            return False
+        if self._report_next_reading:
```

Comment:
I've updated the log messages to announce when it first isn't able to read, and that it will announce when reading resumes. This will typically happen at the very beginning because the first reading will often fail because we start reading at some arbitrary point in the input.

And I've updated the docstring also.